### DISCUSSION TITLE
Using ES Modules with babel-standalone

### DISCUSSION THREAD

**Discussion by georgekamar** [babel/babel#12059] [Q&A] [answered]
Quoting babel docs https://babeljs.io/docs/en/babel-standalone#usage :

> "If you want to use your browser's native support for ES Modules, you'd normally need to set a type="module" attribute on your script tag. With @babel/standalone, set a data-type="module" attribute instead" 

For some reason though, when including my main index.js file (which imports other js / jsx files using import), it seems like babel is converting my import statements to require statements because I get the ReferenceError: require is not defined.

The only way around this I found was to use the transform-modules-umd plugin and include all my js files as scripts. Not sure if this is a bug where data-type="module" doesn't work or if I'm missing something.

These are my scripts tags in index.html

````
    <script src="https://unpkg.com/react@16/umd/react.development.js" crossorigin></script>

    <script src="https://unpkg.com/react-dom@16/umd/react-dom.development.js" crossorigin></script>

    <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>

    <script src="index.js" type="text/babel" data-type="module"></script>
````

Thanks for the help

**Comment by hzoo** [comment, marked as answer]
I believe this is because the default plugins/presets change the code to use require.

https://github.com/babel/babel/blob/1334bc99e05ef863aa372cc1b692cffedb5d50ec/packages/babel-standalone/src/transformScriptTags.js#L39-L44

I assume you can make your own preset/config https://babeljs.io/docs/en/babel-standalone#custom-presets-passing-options-to-built-in-presetsplugins and specify what you need, https://github.com/babel/babel/pull/11466, like preset-env with modules: false https://babeljs.io/docs/en/babel-preset-env#modules

**Reply by georgekamar** [reply]
Thank you for clarifying, I didn't know there were default presets with babel standalone. Using this custom preset (with modules false or even without the env / es2015 preset at all), I noticed that babel standalone does not process the ES modules that are imported. Is that a missing feature in babel standalone ?

**Reply by Amareis** [reply]
Nope, it's just not in babel scope. You need to include all your scripts to document or intercept requests. Little trick - for development purposes you can use service worker with babel-standalone, which will transform all scripts after fetching. You only need to add some bootstrap code, which register service worker and then execute your scripts.

**Comment by samuelcolvin** [comment]
For anyone else looking at this here's my service worker as per @Amareis's suggestion:

`service-worker.js`:

```js
// transpile JSX using babel so it can be run in the browser, you don't need to edit this
// if you do edit this file, call `await window.serviceWorkerRegistration.unregister()` in the browser to update it

self.addEventListener('install', e => e.waitUntil(getBabel()))
self.addEventListener('fetch', e => e.respondWith(handleRequest(e.request)))

async function getBabel() {
  const r = await fetch('https://unpkg.com/@babel/standalone/babel.min.js')
  eval(await r.text())
}

async function handleRequest(request) {
  const url = new URL(request.url)
  const r = await fetch(request)
  if (r.status === 200 & url.host === location.host && url.pathname.endsWith('.js')) {
    const jsx = await r.text()
    const js = Babel.transform(jsx, {presets: ['react']}).code
    return new Response(js, r)
  } else {
    return r
  }
}
```

And to install it:

```html
<script>
  function onError(err) {
    console.error('Error registering service-worker:', err)
    document.getElementById('root').innerText = err.toString()
  }
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.register('/service-worker.js', {scope: '/'})
      .then(registration => {
        window.serviceWorkerRegistration = registration
        // use `await window.serviceWorkerRegistration.unregister()` to unregister the service worker
      })
      .catch(onError)
  } else {
    onError('Browser does not support service workers :-(')
  }
</script>
```

---

This is working for me (except for the first page load where the service worker isn't registered and everything fails) but I'd love to hear if there's a cleaner/ less hacky solution?

**Reply by pmonty** [reply]
@samuelcolvin trying to understand your example. If I have multiple packages to load for example like react-hook-form and redux that my jsx script in my app uses do I need mutliple methods like `getBabel()` and then how do I transpile them all together?

Second part after that is the resultant `.code` can that be then added to my custom jsx that is fetched from a db and then transformed again so it can run in babel/stand-alone. Because my app is currently setup similar to this codesandbox https://codesandbox.io/s/jgm04?file=/src/editor.js:1035-1054 so the service workers transpiled result will be needed as part of the jsx i fetch i believe?
